### PR TITLE
optimize balloc group selection and RAID alignment

### PR DESCRIPTION
**Pull request by jadhavik1**
select group which has enough maxchunk before taking grp lock

### CODE REVIEW

**Comment by jadhavik1** [comment]
With this change we got 40% improvement in write, not more impact on read

File: balloc/balloc.c [hunk 1]
```diff
@@ -2283,9 +2388,9 @@ static int balloc_try_best_found(struct balloc_allocation_context *bac,
 
 		balloc_debug_dump_extent(__func__, &bac->bac_final);
 		M0_ASSERT(is_extent_free(grp, &bac->bac_final, alloc_flag,
-					 &cur));
+					 &cur, &next));
```
Comment (huanghua78) [inline balloc/balloc.c:2391]: M0_ASSERT(), and other ASSERT macro, should have no side-effect.
The M0_ASSERT() did not change things in grp, or bac, but it sill changes the cur, and next.
So, it's better to save the result from is_extent_free() and then assert the result.

File: balloc/balloc.c [hunk 2]
```diff
@@ -2463,30 +2569,71 @@ balloc_regular_allocator(struct balloc_allocation_context *bac)
...
+		 */
+		bac->bac_best.e_start = zp->bzp_curchunk_ext.e_start;
+		bac->bac_best.e_end = zp->bzp_curchunk_ext.e_end;
+		bac->bac_final.e_start = bac->bac_best.e_start;
+		bac->bac_final.e_end = bac->bac_final.e_start +
+				min_check(m0_ext_length(&bac->bac_best), len);
+		bac->bac_status = M0_BALLOC_AC_FOUND;
+		rc = 0;
+		if (0)
+		M0_LOG(M0_ERROR, "<<< g_start:%"PRIu64"  b_start:%"PRIu64" f_start=:%"PRIu64" len=%"PRIu64" culen=%"PRIu64" mlen=%"PRIu64" mstart:%"PRIu64" gr=%"PRIu64"",  bac->bac_goal.e_start, bac->bac_best.e_start,  bac->bac_final.e_start, len, m0_ext_length(&zp->bzp_curchunk_ext), zp->bzp_maxchunk, zp->bzp_maxchunk_ext.e_start, grp->bgi_groupno);
```
Comment (huanghua78) [inline balloc/balloc.c:2604]: Indentation.

File: balloc/balloc.c [hunk 3]
```diff
@@ -2463,30 +2569,71 @@ balloc_regular_allocator(struct balloc_allocation_context *bac)
...
+			rc = 0;
+		} else {
+			if (cr == 0 ||
+				(cr == 1 && len == bac->bac_ctxt->cb_sb.bsb_stripe_size))
+				rc = balloc_simple_scan_group(bac, grp, alloc_type);
+			else
+				rc = balloc_wild_scan_group(bac, grp, alloc_type);
+		}
+	}
+	//M0_LOG(M0_ERROR, "** best.e_start:%"PRIu64" bac->bac_final.e_start=:%"PRIu64" len=%"PRIu64"",  bac->bac_best.e_start,  bac->bac_final.e_start, len);
```
Comment (huanghua78) [inline balloc/balloc.c:2627]: Indentation, and no C++ style comments.

**Comment by huanghua78** [comment]
> With this change we got 40% improvement in write, not more impact on read

Fantastic.

How did you test it?
What is the load?
Please share the details.

**Comment by jadhavik1** [comment]
Test details available at https://seagatetechnology.sharepoint.com/:x:/r/sites/gteamdrv1/tdrive1224/_layouts/15/Doc.aspx?sourcedoc=%7B2380064e-7ae9-4d7d-a35d-363f8e4724c9%7D&action=edit&activeCell=%27balloc_exp_256K_128MB%27!C1&wdrcid=292a446d-6cc4-494a-b42a-b1626eed81eb&wdrldc=1

**Comment by huanghua78** [comment]
I will close this. @shashank-parulekar , @sg-shankar do you plan to have someone else to continue the work?

**Pull request closed without merging**